File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/PinResetViewController.m
//
//  PinResetViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship  on 1/19/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "PinResetViewController.h"

@interface PinResetViewController ()
@property(nonatomic,strong)Utils * utils;
@end

@implementation PinResetViewController
#pragma mark viewDidLoad

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.enterPin.keyboardType=UIKeyboardTypeNumberPad;
    self.enterPin.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* enterPinNumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    enterPinNumberToolbar.barStyle = UIBarStyleBlackTranslucent;
    enterPinNumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinCancelNumberPad)],
                            [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                            [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(enterPinDoneWithNumberPad)]];
    [enterPinNumberToolbar sizeToFit];
    self.enterPin.inputAccessoryView = enterPinNumberToolbar;
    
    self.confirmPin.keyboardType=UIKeyboardTypeNumberPad;
    self.confirmPin.keyboardAppearance = UIKeyboardAppearanceLight;
    UIToolbar* confirmPinNumbertoolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
    confirmPinNumbertoolbar.barStyle = UIBarStyleBlackTranslucent;
    confirmPinNumbertoolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(confirmPincancelNumberPad)],
                             [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                             [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(confirmPinDoneWithNumberPad)]];
    [confirmPinNumbertoolbar sizeToFit];
    self.confirmPin.inputAccessoryView = confirmPinNumbertoolbar;
    
    self.title=@"Reset Pin";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    
    self.resetPinButton.layer.cornerRadius=22.0;
    [self.resetPinButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.resetPinButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.resetPinButton.layer setShadowOpacity:0.5];
   // self.resetPinButton.layer.borderWidth=1.0;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}

#pragma mark viewWillAppear

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    self.enterPin.text=@"";
    self.confirmPin.text=@"";
}
#pragma mark EIDA Toolkit calling functions

- (IBAction)resetPinButtonAction:(id)sender {

  [self.enterPin resignFirstResponder];
  [self.confirmPin resignFirstResponder];
  if ([self.enterPin.text length]==0) {
     [AlertView showAlertTitle:ALERT withMessage:ENTER_PIN onView:self];
  }// if
  else if ([self.confirmPin.text length]==0) {
     [AlertView showAlertTitle:ALERT withMessage:CONFIRM_PIN onView:self];
    }// else if
  else if (![self.enterPin.text isEqualToString:self.confirmPin.text]) {
     [AlertView showAlertTitle:ALERT withMessage:MISMATCH_PIN onView:self];
        self.confirmPin.text=@"";
        self.enterPin.text=@"";
  }// else if
  else {
     if ([self.enterPin.text length]<4 && [self.confirmPin.text length]<4) {
          [AlertView showAlertTitle:ALERT withMessage:MIN_PIN onView:self];
            self.confirmPin.text=@"";
            self.enterPin.text=@"";
     }// if
     else if ([self.enterPin.text length]>16 && [self.confirmPin.text length]>16) {
          [AlertView showAlertTitle:ALERT withMessage:MAX_PIN onView:self];
            self.confirmPin.text=@"";
            self.enterPin.text=@"";
     }// else if
    else {
    [self.utils ShowProgressBar:RESETPIN andView:self.view];
    NSString *requestId=[Utils generateSecureKey];
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
    @try {
        CardReader *cardreader  = [self.model.cardReaderSharedArray objectAtIndex:0];
        FingerData *selectedFingerData =[self.model.selectedFingerSharedArray objectAtIndex:0];
        DataProtectionKey *data_protection_key=self.model.dataprotectedShared;

        NSString *request_handle =  [cardreader prepareRequest:requestId];
        if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
            [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
        }//if
        else {
            NSString *encodePin=[Utils setEncrytion:request_handle data:self.enterPin.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
            if (encodePin.length==0 || [encodePin isEqualToString:@""] || [encodePin isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:ENCODE_PIN_EMPTY onView:self];
            }//if
            else {
               ToolkitResponse *response = [cardreader resetPin:encodePin fingerData:selectedFingerData sensorTimeout:SENSOR_TIME_OUT];
                dispatch_async(dispatch_get_main_queue(), ^{
                    NSString *xmlString =[response getXmlString];
                if (![xmlString isEqualToString:@""] && xmlString.length>0) {
                    NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                    if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        NSString *xmlString =[response getXmlString];
                        [AlertView showAlertTitle:ALERT withMessage:[NSString stringWithFormat:@"%@\n\n%@",PINREST_SUCCESS,xmlString] onView:self];
                    }//else
                }//if
                else {
                   ResponseStatus *status = [response getResponseStatus];
                    if (status==0) {
                        [AlertView showAlertTitle:ALERT withMessage:PINREST_SUCCESS onView:self];
                    }//if
                }//else
                });// update UI main queue
            }//else
          }//else
      }//try
        @catch (NSException *exception) {
            NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
            [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
         }// catch
        @finally {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.confirmPin.text=@"";
                self.enterPin.text=@"";
                [self.utils DismissProgressBar];
             });// update UI main queue
            }// finally
         });// background queue
      }//else
   } //else
}
#pragma mark Updating UI Methods

-(void)enterPinCancelNumberPad {
    [self.enterPin resignFirstResponder];
    self.enterPin.text = @"";
}
-(void)enterPinDoneWithNumberPad {
    [self.enterPin resignFirstResponder];
}
-(void)confirmPincancelNumberPad {
    [self.confirmPin resignFirstResponder];
    self.confirmPin.text = @"";
}
-(void)confirmPinDoneWithNumberPad {
    [self.confirmPin resignFirstResponder];
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */
@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/PinResetViewController.swift
//
//  PinResetViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 25/04/22.
//

import UIKit

class PinResetViewController: UIViewController {
    
    @IBOutlet weak var enterPin: UITextField!
    @IBOutlet weak var confirmPin: UITextField!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)

        self.enterPin.keyboardType = .numberPad
        self.enterPin.keyboardAppearance = .light
        let enterPinNumberToolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        enterPinNumberToolbar.barStyle = .blackTranslucent
        enterPinNumberToolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(enterPinCancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(enterPinDoneWithNumberPad))]
        enterPinNumberToolbar.sizeToFit()
        self.enterPin.inputAccessoryView=enterPinNumberToolbar
        
        self.confirmPin.keyboardType = .numberPad
        self.confirmPin.keyboardAppearance = .light
        let confirmPinNumbertoolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        confirmPinNumbertoolbar.barStyle = .blackTranslucent
        confirmPinNumbertoolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(confirmPincancelNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(confirmPinDoneWithNumberPad))]
        self.confirmPin.inputAccessoryView = confirmPinNumbertoolbar
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        
        self.enterPin.text=""
        self.confirmPin.text=""
    }
    
    @IBAction func btn_LoadFingerData(_ sender: Any) {
        let vc = storyboard?.instantiateViewController(withIdentifier: "FingerDataViewController") as! FingerDataViewController
        present(vc, animated: true, completion: nil)
    }
    
    @IBAction func btn_PinReset(_ sender: Any) {
       
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        else if cardreader == nil {
            self.view.makeToast(Constants.cardreaderNotConnect)
            return
        }//else if
        else if selectedFingerData == nil {
            self.view.makeToast(Constants.fingerDataNotFound)
            return
        }//else if
        
        self.enterPin.resignFirstResponder()
        self.confirmPin.resignFirstResponder()
        
        if self.enterPin.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterPin, onView: self)
        }// if
        else if self.confirmPin.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.ConfirmPin, onView: self)
        }// else if
        else if !(self.enterPin.text == self.confirmPin.text) {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MismatchPin, onView: self)
            self.enterPin.text=""
            self.confirmPin.text=""
        }// else if
        else {
            if (self.enterPin.text?.count)!<4 && (self.confirmPin.text?.count)!<4 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MinPin, onView: self)
                self.enterPin.text=""
                self.confirmPin.text=""
            }// if
            else if (self.enterPin.text?.count)!>16 && (self.confirmPin.text?.count)!>16 {
                AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.MaxPin, onView: self)
                self.enterPin.text=""
                self.confirmPin.text=""
            }// else if
            else {
                utils.showProgressBar(Constants.ResetPin, andView: self.view)
                let requestId = Utils.generateSecureKey()
                DispatchQueue.global(qos: .background).async {
                    do {
                        let request_handle = try cardreader?.prepareRequest(requestId!)
                        if (request_handle?.isEmpty == true) || (request_handle == "") || (request_handle == "(null)") {
                            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
                        }//if
                        else {
                            let encodePin:String? = Utils.setEncrytion(request_handle, data: self.enterPin.text!,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                            if (encodePin?.isEmpty)! || (encodePin == "") || (encodePin == "(null)") {
                                AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Pin_Empty, onView: self)
                            }//if
                            else {
                                response = try cardreader?.resetPin(encodePin!, fingerData: selectedFingerData, sensorTimeout: Int32(Constants.SensorTimeOut))
                                DispatchQueue.main.async {
                                    let xmlString = response.getXmlString()
                                    if (xmlString != "") && ((xmlString.count ) > 0) {
                                        let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
                                        if (errorMsg != "") && (errorMsg.count ) > 0 {
                                            AlertView.showAlertTitle(Constants.Alert, withMessage: errorMsg, onView: self)
                                        }//if
                                        else {
                                            let xmlString = response.getXmlString()
                                            AlertView.showAlertTitle(Constants.Alert, withMessage: "\(Constants.PinResetSuccessStatus)\n\n \(xmlString)", onView: self)
                                        }//else
                                    }//if
                                    else {
                                        let status = response.getResponseStatus()
                                        if status == "0" {
                                            AlertView.showAlertTitle(Constants.Alert, withMessage: "\((Constants.PinResetSuccessStatus))",onView: self)
                                        }//if
                                    }//else
                                }// update UI main queue
                            }//else
                        }//else
                        DispatchQueue.main.async {
                            utils.dismissProgressBar()
                            self.enterPin.text=""
                            self.confirmPin.text=""
                        }// update UI main queue
                    }//do
                    catch let error as NSError{
                        DispatchQueue.main.async {
                            let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                            AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                            utils.dismissProgressBar()
                            self.enterPin.text=""
                            self.confirmPin.text=""
                        }// update UI main queue
                    }//catch
                }//background queue
            }//else
        }//else
    }
    
    @objc func enterPinCancelNumberPad() {
        self.enterPin.resignFirstResponder()
        self.enterPin.text=""
    }
    
    @objc func enterPinDoneWithNumberPad() {
        self.enterPin.resignFirstResponder()
    }
    
    @objc func confirmPincancelNumberPad() {
        self.confirmPin.resignFirstResponder()
        self.confirmPin.text=""
    }
    
    @objc func confirmPinDoneWithNumberPad() {
        self.confirmPin.resignFirstResponder()
    }

    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }

}
